### PR TITLE
Story review platform/create comment

### PR DESCRIPTION
**Pull request by lynn-kim**
## Notion ticket link

[Create new / reply to comment](https://www.notion.so/uwblueprintexecs/Create-new-or-reply-to-comment-27fecc42fa6d4e05a91eeddee1e12245)

## Implementation description
* added a graphql mutation to create a new comment 
* created a comment request and response DTO

## Steps to test
1. make sure you have the comments table in your DB and have populated the `story_translation_contents` table (if not, then set `erase_db_and_sync = True`) 
2. go to http://localhost:5000/graphql and test mutation ([sample query](http://localhost:5000/graphql?variables=null&operationName=createComment&query=mutation%20createComment%7B%0A%20createComment(commentData%3A%7Btime%3A%20123456%2C%20userId%3A%201%2C%20content%3A%22hello%22%2C%20storyTranslationContentId%3A%202%2C%20commentIndex%3A0%2C%20resolved%3Afalse%7D)%20%7B%0A%20%20ok%0A%20%20comment%7B%0A%20%20%20%20content%0A%20%20%20%20time%0A%20%20%20%20userId%0A%20%20%20%20id%0A%20%20%20%20storyTranslationContentId%20%0A%20%20%7D%0A%7D%0A%7D ))
3. try with different values

## What should reviewers focus on?
* `backend/python/app/graphql/mutations/comment_mutation.py`
* `backend/python/app/graphql/types/comment_type.py`
* should comment time be stored as DateTime? 

## Checklist
- [ x ] My PR name is descriptive and in imperative tense
- [ x ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ x ] For backend changes, I have run the appropriate linters:  `docker exec -it <backend-container-id> /bin/bash -c "black . && isort --profile black ."`
- [ x ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by gaoxk — CHANGES_REQUESTED** [review]
Thanks for making the comment service! Forgot to scope that out. 

I like that you linked the query in `Steps to Test` :)

I'm going to backtrack on my initial thoughts on `commentIndex `- after thinking about it more, there's a possibility of bad actors trying to insert the incorrect order and force the frontend into a bad state. Can you remove `commentindex` as a parameter on the mutation and modify the comment service to commit the correct `commentindex` to sqlalchemy?

File: backend/python/app/models/comment.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from sqlalchemy import inspect
-from sqlalchemy.orm.properties import ColumnProperty
 from sqlalchemy.dialects.mysql import LONGTEXT
+from sqlalchemy.orm.properties import ColumnProperty
```
[the remark this answers is not in the document]
Comment (lynn-kim) [reply backend/python/app/models/comment.py:3]: Yep

File: backend/python/app/services/implementations/story_service.py [hunk 2]
```diff
@@ -156,9 +156,6 @@ def get_story_translation(self, id):
             raise error
 
     def assign_user_as_reviewer(self, user, story_translation):
-        print(story_translation)
```
Comment (gaoxk) [inline backend/python/app/services/implementations/story_service.py:159]: LMFAO

File: backend/python/app/graphql/types/comment_type.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import graphene
+
+
+class CreateCommentDTO(graphene.InputObjectType):
+    story_translation_content_id = graphene.Int(required=True)
+    user_id = graphene.Int(required=True)
+    time = graphene.Int(required=True)
```
Comment (gaoxk) [inline backend/python/app/graphql/types/comment_type.py:7]: can we instead insert time inside the story service? i.e. not allow the user to pass in the time

File: backend/python/app/graphql/types/comment_type.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    content = graphene.String(required=True)
+    resolved = graphene.Boolean(required=True)
```
Comment (gaoxk) [inline backend/python/app/graphql/types/comment_type.py:9]: let's not take in `resolved` as a param, comments should be created unresolved

File: backend/python/app/services/implementations/comment_service.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            new_comment = Comment(**comment)
+            comment_index = 0
+
+            comments = Comment.query.filter_by(
+                story_translation_content_id=comment["story_translation_content_id"]
+            )
+
+            if comments.first() is not None:
+                comment_index = (
+                    comments.order_by(Comment.comment_index)[-1].comment_index + 1
```
Comment (gaoxk) [inline backend/python/app/services/implementations/comment_service.py:23]: nice! Is there something like `.max()` in sqlalchemy or mysql? I wonder if there's something faster than sorting.

**Review by gaoxk — CHANGES_REQUESTED** [review]

File: backend/python/app/services/implementations/comment_service.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            new_comment.time = datetime.utcnow()
+            new_comment.resolved = False
+            comment_index = 0
+
+            max = db.session.query(func.max(Comment.comment_index)).filter(
+                Comment.story_translation_content_id
+                == comment["story_translation_content_id"]
+            )
+            last_comment = (
+                db.session.query(Comment.comment_index.label("comment_index"))
```
Comment (gaoxk) [inline backend/python/app/services/implementations/comment_service.py:27]: `Comment.comment_index.label("comment_index")` is this necessary? I think this is relabeling `comment_index` to `comment_index`

File: backend/python/app/services/implementations/comment_service.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                .filter(
+                    Comment.story_translation_content_id
+                    == comment["story_translation_content_id"],
+                    Comment.comment_index == max,
+                )
+                .first()
+            )
+
+            if last_comment is not None:
```
Comment (gaoxk) [inline backend/python/app/services/implementations/comment_service.py:36]: is querying for `last_comment` neccessary? could line 37 instead be `comment_index = max + 1`?

File: backend/python/app/services/implementations/comment_service.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        self.logger = logger
+
+    def create_comment(self, comment):
+        try:
+            new_comment = Comment(**comment)
+            new_comment.time = datetime.utcnow()
+            new_comment.resolved = False
+            comment_index = 0
+
+            max = db.session.query(func.max(Comment.comment_index)).filter(
```
Comment (gaoxk) [inline backend/python/app/services/implementations/comment_service.py:22]: out of curiosity: is a pattern like `Comment.query(...).filter(...)` possible in conjunction with using the `func.max`?

Comment (lynn-kim) [reply backend/python/app/services/implementations/comment_service.py]: I tried earlier but ran into an issue with aggregate functions ([here](https://stackoverflow.com/questions/41887460/select-list-is-not-in-group-by-clause-and-contains-nonaggregated-column-inc/41887524)), but wasn't sure if changing the sql mode variable was the best idea

Comment (gaoxk) [reply backend/python/app/services/implementations/comment_service.py]: Interesting, good to know

**Comment by gaoxk** [comment]
Getting an error with the mutation
```
mutation createComment{
 createComment(commentData:{userId: 1, content:"hello", storyTranslationContentId: 2}) {
  ok
  comment{
    content
    time
    userId
    id
    storyTranslationContentId 
  }
}
}
```
```
{
  "errors": [
    {
      "message": "(MySQLdb._exceptions.DataError) (1265, \"Data truncated for column 'time' at row 1\")\n[SQL: INSERT INTO comments (story_translation_content_id, user_id, comment_index, `time`, resolved, content) VALUES (%s, %s, %s, %s, %s, %s)]\n[parameters: (2, 1, 0, datetime.datetime(2021, 7, 21, 19, 8, 28, 17668), 0, 'hello')]\n(Background on this error at: http://sqlalche.me/e/13/9h9h)",
      "locations": [
        {
          "line": 2,
          "column": 2
        }
      ],
      "path": [
        "createComment"
      ]
    }
  ],
  "data": {
    "createComment": null
  }
}
```

**Review by gaoxk — APPROVED** [review]

**Pull request merged**